### PR TITLE
add inline comments to collections

### PR DESCRIPTION
**Pull request by amgando**

### CODE REVIEW

**Comment by mikedotexe** [comment]
Whoops, I think you need to run `yarn`, @amgando. Looks like there was a change to `package.json` but now the lockfile is off ❤️

**Comment by amgando** [comment]
> Whoops, I think you need to run `yarn`, @amgando. Looks like there was a change to `package.json` but now the lockfile is off ❤️

just committed `yarn.lock`, thanks for the hint

**Comment by amgando** [comment]
also looks like we're no longer compatible w node `10.x` and have set `12.x` as min version?

image

**Comment by mikedotexe** [comment]
> also looks like we're no longer compatible w node `10.x` and have set `12.x` as min version?
> 
> image

checks out. looks like Willem bumped the version in `package.json` in the `engines` key yesterday. I'd try rebasing. Not sure why only Pika is failing, though.

**Comment by willemneal** [comment]
I left an issue with the Pika folks.  For now I disabled it, but it seems it still lingers here.  So you can just ignore it.

File: assembly/runtime/collections/index.ts [hunk 1]
```diff
@@ -1,15 +1,40 @@
...
+ * All contract data is stored in the same key-value data store on the blockchain
+ * (called `Storage` and imported from `near-sdk-as`) with a few convenience
+ * methods for reading, writing, searching and deleting keys and value.
+ *
+ * We also provide a few collections for convenience including
+ * - `PersistentVector`
+ * - `PersistentMap`
+ * - `PersistentDeque`
+ *
+ * These collections wrap the `Storage` class to mimic a Map, Vector (aka. Array) and Deque.
```
Comment (willemneal) [inline assembly/runtime/collections/index.ts:13]: aka. -> aka 
https://writingexplained.org/aka-or-aka

File: assembly/runtime/collections/index.ts [hunk 2]
```diff
@@ -1,15 +1,40 @@
...
+ * And of course you can use these as examples as inspiration for your own custom data structures.
+ *
+ * All of these collections read and write from `Storage` abstracting away a lot of what you might
+ * want to add to the `Storage` object.
+ *
+ * IMPORTANT NOTE:
+ * Since all data stored on the blockchain is kept in a single key-value store under your account,
```
Comment (willemneal) [inline assembly/runtime/collections/index.ts:20]: Isn't it under the contract's account?

File: assembly/runtime/collections/persistentDeque.ts [hunk 3]
```diff
@@ -2,20 +2,49 @@ import { collections } from "../collections";
...
+ * ```
+ *
+ * To use the deque
+ *
+ * ```ts
+ * dq.pushFront(value)
+ * dq.popBack()
+ * ```
+ *
+ * IMPORTANT NOTE:
```
Comment (willemneal) [inline assembly/runtime/collections/persistentDeque.ts:22]: Maybe instead have a link to the the index?

File: assembly/runtime/collections/persistentDeque.ts [hunk 4]
```diff
@@ -71,21 +100,51 @@ export class PersistentDeque<T> {
...
+   *
+   * ```ts
+   * let dq = PersistentDeque<string>("q")
+   *
+   * dq.length // 0
+   * dq.pushFront("hello world")
+   * dq.length // 1
+   * ```
+   *
+   * @returns The length of the deque.
```
Comment (willemneal) [inline assembly/runtime/collections/persistentDeque.ts:131]: Seems redundant with first line

File: assembly/runtime/collections/persistentDeque.ts [hunk 5]
```diff
@@ -71,21 +100,51 @@ export class PersistentDeque<T> {
...
+   */
   get length(): i32 {
     return this.backIndex - this.frontIndex + 1;
   }
 
   /**
+   * Returns a boolean indicating whether the deque is empty
```
Comment (willemneal) [inline assembly/runtime/collections/persistentDeque.ts:138]: Again seems redundant.

File: assembly/runtime/collections/index.ts (hunk 1 above)
Comment (amgando) [reply assembly/runtime/collections/index.ts:13]: unfortunately they provide zero examples of using `aka` and 3 examples of using `a.k.a.`

File: assembly/runtime/collections/index.ts (hunk 2 above)
Comment (amgando) [reply assembly/runtime/collections/index.ts:20]: yes for sure, will update

File: assembly/runtime/collections/persistentDeque.ts (hunk 3 above)
Comment (amgando) [reply assembly/runtime/collections/persistentDeque.ts:22]: for important notes like this i'm in favor of repetition.  it's worse for maintenance but this is unlikely to change without a major overhaul to the design of storage so very low frequency change or no change

File: assembly/runtime/collections/persistentDeque.ts (hunk 4 above)
Comment (amgando) [reply assembly/runtime/collections/persistentDeque.ts:131]: yes, lots of this kind of thing here.  what would be better?

File: assembly/runtime/collections/persistentDeque.ts (hunk 5 above)
Comment (amgando) [reply assembly/runtime/collections/persistentDeque.ts:138]: yep, same as above. what's the right way to handle this?  bc this happens all over the place

File: assembly/runtime/collections/index.ts (hunk 2 above)
Comment (amgando) [reply assembly/runtime/collections/index.ts]: ✅

File: assembly/runtime/collections/persistentDeque.ts (hunk 5 above)
Comment (willemneal) [reply assembly/runtime/collections/persistentDeque.ts:138]: I think change to sentences like: "Indicates whether the deque is empty".

**Pull request merged**